File: packages/test-plugin/ios/Sources/TestPlugin/TestError.swift
import Foundation

/**
 Native error model for the Test plugin (iOS).

 This enum represents ALL possible error categories
 that the native implementation may produce.

 IMPORTANT:
 - Must NOT reference Capacitor
 - Must NOT reference JavaScript or TestErrorCode
 - Mapping to JS error codes happens ONLY in the Plugin layer
 */
enum TestError: Error {
    case unavailable(String)
    case permissionDenied(String)
    case initFailed(String)
    case unknownType(String)

    /// Human-readable error message
    var message: String {
        switch self {
        case .unavailable(let msg),
             .permissionDenied(let msg),
             .initFailed(let msg),
             .unknownType(let msg):
            return msg
        }
    }
}

File: packages/test-plugin/android/src/main/java/io/capkit/test/TestError.kt
package io.capkit.test

/**
 * Native error model for the Test plugin (Android).
 *
 * IMPORTANT:
 * - Extends Throwable to allow usage with Result.failure(...)
 * - Must NOT reference Capacitor or JavaScript
 * - Mapping to JS error codes happens ONLY in the Plugin layer
 */
sealed class TestError(
  message: String,
) : Throwable(message) {
  /** Feature or capability is not available on this device */
  class Unavailable(message: String) : TestError(message)

  /** Permission was denied by the user or system */
  class PermissionDenied(message: String) : TestError(message)

  /** Plugin failed to initialize correctly */
  class InitFailed(message: String) : TestError(message)

  /** Unsupported or unknown type/value was provided */
  class UnknownType(message: String) : TestError(message)
}
